### ISSUE TITLE
jpeg image data from fbfileconsumer's consumeData:(NSData*) data

### ISSUE THREAD

**Issue by aravindgs** [facebook/idb#433]
Hi 

I wanted to convert the data received in the FBFileConsumer's consumeData:  method into a stream of jpegs. From my understanding of the code the data is CVPixelBufferRef converted to NSData.

I tried regenerating the CVPixelBufferRef and obtaining the NSImage and hence getting the jpeg data from NSImage. Unfortunately this has not worked. Could you please help me out with this? Adding below the code i tried (the code is in swift).

`
        var pixelBuffer:CVPixelBuffer? = nil

        let result = CVPixelBufferCreate(kCFAllocatorDefault, 750, 1334, kCVPixelFormatType_32BGRA, nil, &pixelBuffer)
        
        if result != kCVReturnSuccess {
            print("pixel buffer create success")
            return
        }
        CVPixelBufferLockBaseAddress(pixelBuffer!, .init(rawValue: 0))
        let yDestPlane:UnsafeMutableRawPointer? = CVPixelBufferGetBaseAddressOfPlane(pixelBuffer!, 0)
        
        if yDestPlane == nil {
            print("failed to create ydestplane")
            return
        }
        
        let nsData = data as NSData
        let rawPtr = nsData.bytes
        
        memcpy(yDestPlane!, rawPtr, 750*1334*4)
        
        CVPixelBufferUnlockBaseAddress(pixelBuffer!, .init(rawValue: 0))
        kCVPixelFormatType_32BGRA, getbasea, 2208, NULL, NULL, nil, &pixelBuffer)
        
        if #available(OSX 10.12, *) {
            let ciImage = CIImage.init(cvPixelBuffer: pixelBuffer!)
            let tempContext = CIContext.init(options: nil)
            let videoImage = tempContext.createCGImage(ciImage, from: CGRect.init(x: 0, y: 0, width: CVPixelBufferGetWidth(pixelBuffer!), height: CVPixelBufferGetHeight(pixelBuffer!)))
            let imageSize: NSSize = NSMakeSize(750, 1334)
            let nsImageTest = NSImage(cgImage: videoImage!, size: imageSize)
            if let bits = nsImageTest.representations.first as? NSBitmapImageRep {
                let jpegFinalData = bits.representation(using: .JPEG, properties: [:])
                if self.isStreaming {
                    var simIndex:Int?
                    for i in 0...(AUConnectionSimulatorMap.instance.simConnectionMap.count-1) {
                        if sim.udid == AUConnectionSimulatorMap.instance.simConnectionMap[i].sim.sim.udid {
                            simIndex = i
                            break
                        }
                    }
                    var finalData:Data = Data()
                    let finalDict = ["data":["type":"onScreenFrame","value":jpegFinalData!]] as Dictionary<String,Any>
                    try! finalData.pack(finalDict)
                    AUConnectionSimulatorMap.instance.simConnectionMap[simIndex!].ws?.write(data: finalData)
                }
                
            }
            
        } else {
            // Fallback on earlier versions
        }`

I am running the code on mac os x 10.12 machine

**Comment by aravindgs** [comment]
Solution

```
            var width:size_t?
            var height:size_t?
            let bitsPerComponent:size_t = 8
            var bytesPerRow:size_t?
            var bitmapInfo: UInt32
            if #available(OSX 10.12, *) {
                bitmapInfo = UInt32(CGImageAlphaInfo.premultipliedFirst.rawValue) | UInt32(CGImageByteOrderInfo.order32Little.rawValue)
            } else {
                bitmapInfo = UInt32(CGImageAlphaInfo.premultipliedFirst.rawValue)
            }
            let colorSpace = CGColorSpaceCreateDeviceRGB()
            
            do {
                let streamAttributesFuture = self.bitmapStream?.streamAttributes()
                streamAttributesFuture?.onQueue(.main, notifyOfCompletion: { (streamAttributesFut) in
                    
                })
                try streamAttributesFuture?.await()
                let dict = streamAttributesFuture?.result
                width = (dict?.attributes["width"] as! Int)
                height = (dict?.attributes["height"] as! Int)
                bytesPerRow = (dict?.attributes["row_size"] as! Int)
                
            } catch let frameError{
                print("frame error = \(frameError)")
                return
            }
            
            let rawPtr = (data as NSData).bytes
            let mutableRawPtr = UnsafeMutableRawPointer.init(mutating: rawPtr)
            let context = CGContext.init(data: mutableRawPtr, width: width!, height: height!, bitsPerComponent: bitsPerComponent, bytesPerRow: bytesPerRow!, space: colorSpace, bitmapInfo: bitmapInfo)
            if context == nil {
                return
            }
            
            let imageRef = context?.makeImage()
            
            let scaledWidth = Int(Double(width!) / 1.5)
            let scaledHeight = Int(Double(height!) / 1.5)
            
            let resizeCtx = CGContext.init(data: nil, width: scaledWidth, height: scaledHeight, bitsPerComponent: bitsPerComponent, bytesPerRow: bytesPerRow!, space: colorSpace, bitmapInfo: bitmapInfo)
            resizeCtx?.draw(imageRef!, in: CGRect(x: 0, y: 0, width: scaledWidth, height: scaledHeight))
            
            resizeCtx?.interpolationQuality = .low
            
            let resizedImgRef = resizeCtx?.makeImage()
            let imageRep = NSBitmapImageRep(cgImage: resizedImgRef!)
            
            let finalData = imageRep.representation(using: .jpeg, properties: [NSBitmapImageRep.PropertyKey.compressionFactor : compression])
```

**Comment by lawrencelomax** [comment]
`FBSimulatorImage` also does something similar, though it's not streamed.
